### PR TITLE
#175261605 cancel offer letter

### PR DESCRIPTION
**Pull request by tobieeniafe**
#### What does this PR do?
Modify endpoint to cancel offer letter

#### Description of Task to be completed?
Allows an admin to cancel an offer letter

#### How should this be manually tested?
`PUT /api/v1/offer/cancel`

#### What are the relevant pivotal tracker stories?
#175261605

### CODE REVIEW

File: server/routes/offer.routes.js [hunk 1]
```diff
@@ -84,7 +84,13 @@ router.put('/accept', authenticate, schemaValidation(acceptOfferSchema), OfferCo
  *      '500':
  *       description: Internal server error
  */
-router.put('/cancel', authenticate, schemaValidation(respondOfferSchema), OfferController.cancel);
+router.put(
+  '/cancel',
+  authenticate,
+  isAdmin,
+  schemaValidation(respondOfferSchema),
```
Comment (andela-hpopoola) [inline server/routes/offer.routes.js:91]: The name here is a little confusing, in we change it into something more descriptive. I am afraid that in 2 weeks time, I would have forgotten the purpose of the schema without checking the code 😅 

```suggestion
  schemaValidation(validateOfferIdSchema),
```

File: server/services/offer.service.js [hunk 2]
```diff
@@ -303,6 +304,10 @@ export const acceptOffer = async (offerToAccept) => {
       'You cannot accept offer of another user',
     );
   }
+
+  if (offer[0].status === OFFER_STATUS.CANCELLED) {
```
Comment (andela-hpopoola) [inline server/services/offer.service.js:308]: 🙌 Thank you for adding this.

File: server/services/offer.service.js [hunk 3]
```diff
@@ -357,12 +362,21 @@ export const assignOffer = async (offerId) => {
...
   const offer = await getOfferById(offerId).catch((error) => {
     throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
   });
 
+  if (offer.vendorId.toString() !== vendorId.toString()) {
+    throw new ErrorHandler(httpStatus.FORBIDDEN, 'You are not permitted to perform this action');
+  }
+
+  if (offer.status === OFFER_STATUS.ASSIGNED || offer.status === OFFER_STATUS.ALLOCATED) {
+    throw new ErrorHandler(httpStatus.PRECONDITION_FAILED, 'You cannot cancel an accepted offer');
```
Comment (andela-hpopoola) [inline server/services/offer.service.js:375]: This is nice, however, we might need more information from the team if this is the right implementation. Apart from offline communication with the team, no code changes is expected for now.

File: test/services/offer.service.test.js [hunk 4]
```diff
@@ -484,16 +484,16 @@ describe('Offer Service', () => {
 
     context('when all is valid', () => {
       it('returns a valid cancelled offer', async () => {
-        const rejectedOffer = await cancelOffer(offerId);
-        expect(rejectedOffer.status).to.eql('Cancelled');
+        const rejectedOffer = await cancelOffer({ offerId, vendorId: userId });
+        expect(rejectedOffer.status).to.eql(OFFER_STATUS.CANCELLED);
```
Comment (andela-hpopoola) [inline test/services/offer.service.test.js:488]: This looks good but a Senior Software engineer had once advised me that it is not best practices for test to use defined constants. His reason is that the test will still pass if the constant is mistakenly updated e.g if someone updates constant.js file and added an unknown character into this, something like 'Cancelled2', the test will still pass.

I think it is a valid reason, even though I knowingly ignore the past constants you used in previous tests to minimize the number of comments on your PR. Henceforth, lets try to minimize or use the real value instead of constants. 
```suggestion
        expect(rejectedOffer.status).to.eql('Cancelled');
```

File: server/services/offer.service.js [hunk 5]
```diff
@@ -357,12 +362,21 @@ export const assignOffer = async (offerId) => {
   }
 };
 
-export const cancelOffer = async (offerId) => {
+export const cancelOffer = async ({ offerId, vendorId }) => {
   const offer = await getOfferById(offerId).catch((error) => {
     throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
   });
 
+  if (offer.vendorId.toString() !== vendorId.toString()) {
```
Comment (andela-hpopoola) [inline server/services/offer.service.js:370]: A test for this will make so much test. Trying to cancel an offer with a different vendor.

File: server/services/offer.service.js [hunk 6]
```diff
@@ -357,12 +362,21 @@ export const assignOffer = async (offerId) => {
...
+    throw new ErrorHandler(httpStatus.FORBIDDEN, 'You are not permitted to perform this action');
+  }
+
+  if (offer.status === OFFER_STATUS.ASSIGNED || offer.status === OFFER_STATUS.ALLOCATED) {
```
Comment (andela-hpopoola) [inline server/services/offer.service.js:374]: I didn't check other parts of the code not touched by this PR, can you verify that we have tests to confirm that users cannot cancel offers that are assigned or allocated.

File: server/services/offer.service.js [hunk 7]
```diff
@@ -357,12 +362,21 @@ export const assignOffer = async (offerId) => {
...
+    throw new ErrorHandler(httpStatus.PRECONDITION_FAILED, 'You cannot cancel an accepted offer');
+  }
+
   try {
+    await Enquiry.findByIdAndUpdate(offer.enquiryId, { $set: { approved: false } });
```
Comment (andela-hpopoola) [inline server/services/offer.service.js:379]: We also need to ensure we have tests that confirms that this happens.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for your hard work. The implementation looks dope, just need more flavour of tests to ensure that the application is solid. Well done.

**Review by andela-hpopoola — CHANGES_REQUESTED** [review]
Thank you for working on this. We need to add more tests to make this more robust. Cancelling Offer letters is a big deal as we don't want to mistakenly cancel a valid offer letter, so we need to ensure that our test covers all edge cases.

Kindly add the following tests:
1. Ensure that we cannot cannot for ASSIGNED, ALLOCATED
2. Ensure that the vendor can cancel a REJECTED and GENERATED offer
3. We should add a test for CANCELLED offer too but not sure if we should allow a vendor to cancel a CANCELLED offer or not. Leaving this to your discretion, however, a test is needed no matter your decision.
4. Ensure that a user cannot cancel an offer letter.

Thank you.

**Comment by andela-hpopoola** [comment]
Thank you for working on this. It looks like you missed the following tests from the comments here
https://github.com/Highrachy/ballers-api/pull/59#pullrequestreview-510141396

> * Ensure that the vendor can cancel a REJECTED ~and GENERATED~ offer
> * We should add a test for CANCELLED offer too but not sure if we should allow a vendor to cancel a CANCELLED offer or not. Leaving this to your discretion, however, a test is needed no matter your decision.
> * Ensure that a user cannot cancel an offer letter.

Can you please add the tests so that we can proceed with other important tasks.

File: test/controllers/offer.controller.test.js [hunk 8]
```diff
@@ -714,6 +746,21 @@ describe('Offer Controller', () => {
       });
     });
 
+    context('when offer is cancelled by unauthorized vendor', () => {
```
Comment (andela-hpopoola) [inline test/controllers/offer.controller.test.js:749]: This should be 

```
when an offer is canceled by a user
```

File: test/controllers/offer.controller.test.js [hunk 9]
```diff
@@ -759,6 +806,40 @@ describe('Offer Controller', () => {
       });
     });
 
+    context('with valid data & token', () => {
```
Comment (andela-hpopoola) [inline test/controllers/offer.controller.test.js:809]:
```
when an offer has been rejected
```

File: test/controllers/offer.controller.test.js [hunk 10]
```diff
@@ -759,6 +806,40 @@ describe('Offer Controller', () => {
...
+            expect(enquiry.approved).to.be.eql(false);
+            expect(res).to.have.status(200);
+            expect(res.body.success).to.be.eql(true);
+            expect(res.body.message).to.be.eql('Offer cancelled');
+            done();
+          });
+      });
+    });
+
+    context('with valid data & token', () => {
```
Comment (andela-hpopoola) [inline test/controllers/offer.controller.test.js:826]:
```
when an offer has been canceled
```

**Review by andela-hpopoola — APPROVED** [review]
Thank you for adding the final tests. Left a few comments on improving the test description. Feel free to merge after updating them.
